### PR TITLE
mgr/dashboard: cephfs volume creation form

### PR DESCRIPTION
**Pull request by pereman2**
image

image

Fixes: https://tracker.ceph.com/issues/62085

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cloudbehl** [comment]
@pereman2 can you add recording or screenshot to the PR?

**Review by nizamial09 — COMMENTED** [review]
Looks mostly good. 

I identified some issue and I commented here. You can check it.

One of the issue is regarding the no orchestrator error. When you try from a slow environment the form loads a bit weirdly. That code needs to be modified to fix it. Using an async pipe will fix it. I put the code below

https://github.com/ceph/ceph/assets/71764184/8f0e810d-d8de-41ff-8780-4b26f6dddd09

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+          serviceSpec['placement']['hosts'] = values['hosts'];
+        }
+        break;
+      case 'label':
+        serviceSpec['placement']['label'] = values['label'];
+        break;
+    }
+
+    const volumeName = this.form.get('name').value;
+    const self = this;
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.ts:139]: tbh just copy pasted from the service.component XD I thought the same

**Comment by pereman2** [comment]
jenkins test api

**Comment by pereman2** [comment]
jenkins test make check

**Review by aaSharma14 — APPROVED** [review]
Tested locally..LGETM!

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 2]
```diff
@@ -37,6 +38,24 @@ def list(self):
         fsmap = mgr.get("fs_map")
         return fsmap['filesystems']
 
+    def create(self, name: str, service_spec: Dict[str, Any]):
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:41]: Could you please add API tests to the PR?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+describe('CephfsVolumeFormComponent', () => {
+  it('empty', () => {});
+});
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.spec.ts:3]:
```suggestion
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { CephfsVolumeFormComponent } from './cephfs-form.component';


import { configureTestBed } from '~/testing/unit-test-helper';
import { SharedModule } from '~/app/shared/shared.module';
import { ToastrModule } from 'ngx-toastr';
import { ReactiveFormsModule } from '@angular/forms';


describe('CephfsVolumeFormComponent', () => {
  let component: CephfsVolumeFormComponent;
  let fixture: ComponentFixture<CephfsVolumeFormComponent>;

  configureTestBed({
    imports: [
      BrowserAnimationsModule,
      SharedModule,
      HttpClientTestingModule, 
      RouterTestingModule,
      ReactiveFormsModule,
      ToastrModule.forRoot()],
    declarations: [CephfsVolumeFormComponent]
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CephfsVolumeFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });
});

```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.ts [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+  authenticate() {
+    this.permission = this.authStorageService.getPermissions().cephfs;
+    if (
+      !this.permission.read ||
+      (!this.permission.update && this.editing) ||
+      (!this.permission.create && !this.editing)
+    ) {
+      throw new DashboardNotFoundError();
+    }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.ts:88]: not needed!

**Review by nizamial09 — APPROVED** [review]
Thanks @pereman2 for addressing the comments!

File: src/pybind/mgr/dashboard/controllers/cephfs.py [hunk 5]
```diff
@@ -37,6 +38,24 @@ def list(self):
...
+            for host in service_spec['placement']['hosts']:
+                service_spec_str += f'{host},'
+            service_spec_str = service_spec_str[:-1]
+
+        error_code, _, err = mgr.remote('volumes', '_cmd_fs_volume_create', None,
+                                        {'name': name, 'placement': service_spec_str})
+        if error_code != 0:
+            raise RuntimeError(
+                f'Error creating volume {name} with placement {str(service_spec)}: {err}')
+        return 'Volume {name} created successfully'
```
Comment (avanthakkar) [inline src/pybind/mgr/dashboard/controllers/cephfs.py:57]:
```suggestion
        return f'Volume {name} created successfully'
```

**Review by avanthakkar — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.spec.ts (hunk 3 above)
Comment (pereman2) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.spec.ts:3]: doesn't work :(

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  }
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/cephfs/cephfs-form/cephfs-form.component.ts:89]: not needed!
```suggestion

```

**Comment by epuertat** [comment]
Not sure if you already fixed but the notice banner doesn't have proper padding/margin:

image

That's probably not placed inside the inner `div.form-body` container or however is called.

File: src/pybind/mgr/dashboard/controllers/cephfs.py (hunk 2 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/controllers/cephfs.py:42]: Would be good to have some API tests and a frontend unit test for cephfs create service as well.

Comment (pereman2) [reply src/pybind/mgr/dashboard/controllers/cephfs.py:42]: As we agreed on the last meeting let's skip these for now for deadline sake.

**Comment by cloudbehl** [comment]
Is there upgraded screenshot after the padding fix?

**Review by avanthakkar — APPROVED** [review]

**Comment by Pegonzal** [comment]
jenkins test make check

**Comment by pereman2** [comment]
jenkins test make check

**Pull request merged**